### ISSUE TITLE
Too slow when works with large file

### ISSUE THREAD

**Issue by codehz** [peco/peco#454]
# Abstract
run peco with a large text file like more than 10M
At first, it fast as expected, but when I input more text, it will be slow down quickly
If query text' length longer then 7 it will take 1 second to fetch result in my laptop

# Discussion
I think it is reasonable for long time to search at first time, but I think the incremental search shouldn't be too slow(especially when there are only few result)

**Comment by lestrrat** [comment]
Fair enough but at that poiny I think we need to be building an index and that's going to affect 1) initial startup performance, and 2) it's going to take up more than double the memory of your dataset. I'm not enitrely sure peco should be going that route

**Comment by lestrrat** [comment]
If you have 50 results displayed (I assume this is what you mean but "current" result), and you type a new keyword, and only perform queries on those 50, what happens when only, say, 10 out of 50 match, but there are a bunch more in the results that are not being displayed? Do you display just those 10? I assume not, as that would lead to false negatives. So you'd have to keep searching until you fill up the buffer, or you exhaust the targets. Hmmm, meh...

Now, I agree that we could use some optimizations, but unless we do something fundamental about it, the worst case is still O(n), and we're ultimately bound by the search performance when dealing with really huge buffers.

ATM your suggestion seems like a micro-optimization that is only marginally useful.

**Comment by lestrrat** [comment]
mental note: maybe [bleve](http://www.blevesearch.com/docs/Query/)
